#162948858 comment on highlighted text

Georgeygigz opened the pull request:
**What does this PR do?**
- It creates a functionality to comment on a specific highlighted text.

**Description of Task to be completed?**
- Highlighting text to be commented.
- Write tests for the functionalities.

**How should this be manually tested?**
- `git clone https://github.com/andela/ah-django-unchained`
- `cd ah-django-unchained`

- `git checkout fft-highlight-comment-text-162948858`

- `create a virtual env python3 -m venv virtual_env_name`

- `pip install -r requirements.txt`

-  `source .env`

**Start in postman using the following url**
- Navigate to this url on postman 
 `http://127.0.0.1:8000/api/articles/dummy-text/highlight/`
-  In the Json body
``` 
{
	"start_highlighting":0,
	"end_highlighting":8,
	"body":"Remove this line"
} 
```
**PT stories**
- Naviagte to this [link](https://www.pivotaltracker.com/story/show/162948858)

**Screenshots**
![image](https://user-images.githubusercontent.com/24627269/52110101-a28d1d00-2610-11e9-8623-a57be22dee9f.png)

Georgeygigz committed:
feat(Highlight text):comment highlighted text

-Comment highlighted text
-Write tests

[Starts #162948858]

kwameasiago reviewed (approved):
Looks Food

ignatiusukwuoma commented on authors/apps/articles/serializers.py:
```diff
@@ -117,7 +118,8 @@ def to_representation(self, instance):
                     instance=UserProfile.objects.get(user=thread.author)).data,
                 'createdAt': self.format_date(thread.createdAt),
-                'updatedAt': self.format_date(thread.updatedAt)
+                'updatedAt': self.format_date(thread.updatedAt),
```
You shouldn't edit this part

Georgeygigz replied:
Okay noted

ignatiusukwuoma commented on authors/apps/articles/models.py:
```diff
@@ -49,10 +49,13 @@ class Comment(models.Model):
     body = models.TextField(blank=False)
     createdAt = models.DateTimeField(auto_now_add=True)
     updatedAt = models.DateTimeField(auto_now=True)
+    selected_text = models.TextField(null=True)
```
I think this implies that I can only highlight one text per article. That assumption would be wrong, I should be able to highlight several pieces of text in an article but these highlights should only be visible to me. I think this will involve creating another table that will have user_id, article_id, highlight_start and highlight_end

ignatiusukwuoma commented on authors/apps/articles/serializers.py:
```diff
@@ -101,6 +101,7 @@ class CommentSerializer(serializers.ModelSerializer):
     """Serializer for creating a Comment"""
     author = serializers.SerializerMethodField()
     body = serializers.CharField()
+    selected_text = serializers.CharField(read_only=True)
```
I think a different serializer will be created for this purpose... per my comment above.

Georgeygigz replied:
Okay noted

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +393,71 @@ def retrieve(self, request, slug, *args, **kwargs):
+        # serializing article details in order to extract
+        # the body of an article
+        serializer = self.serializer_class(single_article_instance)
+        word = serializer.data['body'].split()
```
I don't expect you to split the words though... because a highlight can be from the middle of the word

Georgeygigz replied:
Okay

ignatiusukwuoma commented:
To get you to understand what is required here, as a user when I highlight a piece of text, the frontend would provide a way for me to comment on it and send the highlight start and end alongside the comment to the backend. Backend then has to save this somewhere that is not directly related to the article but will have a reference to the article so that when the user returns to that article, we can show them their comments.

Georgeygigz commented:
> To get you to understand what is required here, as a user when I highlight a piece of text, the frontend would provide a way for me to comment on it and send the highlight start and end alongside the comment to the backend. Backend then has to save this somewhere that is not directly related to the article but will have a reference to the article so that when the user returns to that article, we can show them their comments.

Okay well noted

Georgeygigz committed:
feat(Refactor highlight implementation):Refactor highlight implementation

-Add model and serializer for the implementation
-Save the index where the highlight starts and also where it ends

[Fixes #162948848]

MaggieChege commented on authors/apps/articles/serializers.py:
```diff
@@ -140,3 +141,32 @@ class DeleteCommentSerializer(serializers.ModelSerializer):
+        return obj.user_id.id
+
+    def get_article(self, obj):
+        return obj.article.slug
```
Add blank line

MaggieChege commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +394,122 @@ def retrieve(self, request, slug, *args, **kwargs):
+        if serializer.is_valid():
+            serializer.save(user_id=request.user,
+                            article=single_article_instance)
+            serialized_details = serializer.data
```
You haven't used serialized_details anywhere.

andrewhingah commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +394,112 @@ def retrieve(self, request, slug, *args, **kwargs):
+                message = {"error": "The comment does not exist"}
+                return Response(message, status.HTTP_404_NOT_FOUND)
+            super().delete(self, request, highlightcomment)
+            message = {"message": "Highlighted Comment deleted successfully"}
```
message should be `Comment deleted successfully` and not `highlighted comment`.

kwameasiago commented:
Add tests to increase your coverage

ignatiusukwuoma commented on authors/apps/articles/models.py:
```diff
@@ -60,3 +60,20 @@ def __str__(self):
+    body = models.TextField()
+    article = models.ForeignKey(Article, related_name='highlight',
+                                on_delete=models.CASCADE, null=True)
+    selected_text = models.TextField(blank=True, null=True)
```
For Textfield and Charfield, don't use null=True so that empty fields are saved with empty strings instead of NULL
https://docs.djangoproject.com/en/2.1/ref/models/fields/#django.db.models.Field.null

Georgeygigz replied:
okay

ignatiusukwuoma commented on authors/apps/articles/serializers.py:
```diff
@@ -119,7 +119,8 @@ def to_representation(self, instance):
                     instance=UserProfile.objects.get(user=thread.author)).data,
                 'createdAt': self.format_date(thread.createdAt),
-                'updatedAt': self.format_date(thread.updatedAt)
+                'updatedAt': self.format_date(thread.updatedAt),
```
Remove comma and the blank space below

Georgeygigz replied:
okay noted

ignatiusukwuoma commented on authors/apps/articles/serializers.py:
```diff
@@ -144,3 +145,31 @@ class Meta:
+    body = serializers.CharField()
+    selected_text = serializers.CharField()
+    start_highlight_position = serializers.CharField()
+    end_highlight_position = serializers.CharField()
```
Aren't these supposed to be IntegerFields ?

Georgeygigz replied:
Yeah they are.

ignatiusukwuoma commented on authors/apps/articles/serializers.py:
```diff
@@ -144,3 +145,31 @@ class Meta:
+        instance.save()
+        return instance
+
+    def get_user_id(self, obj):
```
There's another method with the same name on line 159, I think you should delete that one.

ignatiusukwuoma commented on authors/apps/articles/urls.py:
```diff
@@ -7,8 +7,11 @@
+                    ReadTime, HighlightText,
+                    CreateComment, CommentDetailsView, CommentDelete,
+                    CommentsRetrieveUpdateDestroy, ReadTime,
+                    HighlightText, RetrieveUpdateDeleteComments)
```
Remove duplicate imports

ignatiusukwuoma commented on authors/apps/articles/urls.py:
```diff
@@ -53,4 +56,10 @@
          name="share_twitter"),
+    path('articles/<slug>/highlight/',
+         HighlightText.as_view(), name='high_light'),
+    path('articles/<slug>/highlight/',
```
Remove duplicate url

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -19,10 +19,12 @@
                           RatingSerializer, CommentSerializer,
-                          DeleteCommentSerializer)
+                          DeleteCommentSerializer, HighlightSerializer,
+                          HighlightTextModel)
```
import HighlightTextModel from .models where it was defined

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -475,3 +477,113 @@ def get(self, request, slug):
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+    def select_index_position(self, start_index_position, end_index_position):
```
this can be static function

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -475,3 +477,113 @@ def get(self, request, slug):
+        return index_values
+
+    def get(self, request, slug):
+        """Get all comment for highligted text of a single article."""
```
typo

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -475,3 +477,113 @@ def get(self, request, slug):
+    def get(self, request, slug):
+        """Get all comment for highligted text of a single article."""
+        article = get_object_or_404(Article, slug=slug)
+        highlighted_comment = article.highlight.filter()
```
Shouldn't you be filtering by the user and the article?

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -475,3 +477,113 @@ def get(self, request, slug):
+    def get(self, request, slug, id):
+        article = get_object_or_404(Article, slug=slug)
+        try:
+            highlightcomment = article.highlight.filter(id=id).first().pk
```
separate words with underscores `highlight_comment`

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -475,3 +477,113 @@ def get(self, request, slug):
+        try:
+            highlightcomment = article.highlight.filter(id=id).first().pk
+            return super().get(request, highlightcomment)
+        except:
```
Too broad

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -475,3 +477,113 @@ def get(self, request, slug):
+                message, status.HTTP_200_OK)
+
+    def update(self, request, slug, id):
+        article = Article.objects.filter(slug=slug).first()
```
Can we have a method that does the job of checking if the article and comment exist and can be called on the get, delete and update methods

Georgeygigz committed:
feat(Refactoring code):Refactoring code

- Remove unused imports
- Remove duplicate imports
- Create a single class to check if article and comments exist
- Change start_highlight_position and end_highlight_position to integer fileds
- Removing typos
- Remove duplicates url

[Fixes #162948858]
